ios-sample-app/ios-sample-app/Controllers/IdentityController.m

```diff
@@ -18,25 +22,162 @@ @implementation IdentityController
+        
+        //Logging for testing
+        NSLog(@"CCPA OptOut - Age is below 13 - Tracking Stopped");
+        [Utils displayMessage:[[NSString alloc] initWithFormat:@"CCPA OptOut - Age is below 13 - Tracking Stopped"] withView:self];
+    } else {
+        [Singular resumeAllTracking];
+        [Singular event:@"CCPA_Disabled"];
+        
+        //Logging for testing
+        NSLog(@"CCPA OptOut Disabled - Tracking Started");
+        [Utils displayMessage:[[NSString alloc] initWithFormat:@"CCPA OptOut Disabled - Tracking Started"] withView:self];
+        }
 }
 
 - (IBAction)setCustomUserId:(id)sender {
     NSString* customUserId = self.customUserIdField.text;
     
     if([Utils isEmptyOrNull:customUserId]){
         [Utils displayMessage:@"Please enter a valid Custom User Id" withView:self];
+    } else {
+        // Once set, the Custom User Id will persist between runs until `[Singular unsetCustomUserId]` is called.
+        // This can also be called before SDK init if you want the first session to include the Custom User Id.
+        [Singular setCustomUserId:customUserId];
+        
+        // Logging for Testing
+        NSLog(@"Custom User ID Set %@", customUserId);
+        [Utils displayMessage:[[NSString alloc] initWithFormat:@"Custom User Id set to: %@", customUserId] withView:self];
     }
-    
-    // Once set, the Custom User Id will persist between runs until `[Singular unsetCustomUserId]` is called.
-    // This can also be called before SDK init if you want the first session to include the Custom User Id.
-    [Singular setCustomUserId:customUserId];
-
-    [Utils displayMessage:[[NSString alloc] initWithFormat:@"Custom User Id set to: %@", customUserId] withView:self];
 }
 
 - (IBAction)unsetCustomUserId:(id)sender {
     [Singular unsetCustomUserId];
+    
+    // Logging for Testing
+    NSLog(@"Custom User ID Unset");
     [Utils displayMessage:@"Custom User Id unset" withView:self];
 }
 
+- (void)requestTrackingAuthorization {
+    AppDelegate *DataDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
+    // call requestTrackingAuthorizationWithCompletionHandler from ATTrackingManager to start the user consent process
+    if (@available(iOS 14, *)) {
+        [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
+            // Tracking authorization completed. Start loading ads here.
+            switch (status) {
+                case 0: self.att_state = @"(0) Not Determined";
+                case 1: self.att_state = @"(1) Restricted";
+                case 2: self.att_state = @"(2) Denied";
+                case 3: self.att_state = @"(3) Authorized";
+                }
+            self.s_idfa = [[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString];
+            self.s_idfv = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
+            [self.idfa_value setText:self.s_idfa];
```

Comment:
set the Labels value in main queue. App may crash sometimes.

Reply:
@tapasyat - I am not sure what this means. Can you elaborate?

Reply:
Updating the UI in main thread.
dispatch_async(dispatch_get_main_queue(), ^{
                [self.idfa_value setText:self.s_idfa];
                [self.idfv_value setText:self.s_idfv];
                [self.att_value setText:self.att_state];
                DataDelegate.sharedMessage = [NSString stringWithFormat:@"Sample App Device Info:\n\nAdvertising ID (IDFA): %@ \n\nIDFV: %@",self.s_idfa,self.s_idfv];
            });